### ISSUE TITLE
additionalProperties causes JsonException: 

### ISSUE THREAD

**Issue by ChadNedzlek** [Azure/autorest#2469]
Newtonsoft.Json.JsonException occurred
  HResult=0x80131500
  Message=Invalid extension data attribute on 'MyType'. Member 'AdditionalProperties' type must implement IDictionary<string, JToken>.
  Source=Newtonsoft.Json
  StackTrace:
   at Newtonsoft.Json.Serialization.DefaultContractResolver.<>c.<GetExtensionDataMemberForType>b__34_1(MemberInfo m)

The actual signature of that property is IDictionary<string,string> (which presumably autorest got out of the JsonSchema).

I was using autorest 1.2.2

**Comment by olydis** [comment]
could you please post the corresponding (part of the) OpenAPI definition? the message indicates that it does look like what AutoRest expects it to look like so it is key to helping you.

**Comment by ChadNedzlek** [comment]
The API I'm calling is at https://helix.dot.net/swagger/docs/2017-04-14, in the "OtherEvent" object.  AutoRest creates a typed dictionary, but the serialize doesn't seem to support that... why is autorest generating types that it can't serialize? I looked in the json.net code on github, and it's pretty explicit that [JsonExtensionData] can only appear on an IDictionary<string, JToken>.

**Comment by olydis** [comment]
Indeed! I'm surprised this was not noticed until now since `additionalProperties` is used quite rarely, whenever I saw it in the past, the value was instantiated as `object` which obviously (de)serializes successfully, too.

>  I looked in the json.net code on github, and it's pretty explicit that [JsonExtensionData] can only appear on an IDictionary<string, JToken>.

Maybe the code is "clear", but do you find any pointer like that in their docs? I think it's a pretty reasonable assumption to expect it to work for other types - that would only make it consistent with regular properties (after all you can always have JToken as the "anything" type, but then make it something more specific... why don't they just do the same for JsonExtensionData? oh well).
Patching the serialization runtime is probably the best fix in this case, anything else would require adding a backing dictionary to generated code and then casting on the fly.

**Comment by ChadNedzlek** [comment]
No, the best I can find in their documentation is that all the examples always use JToken, but the exception message itself is quite clear, and I suppose that's a... kind of documentation.  The code that I think is throwing the exception is [here](https://github.com/JamesNK/Newtonsoft.Json/blob/bbbddf0031fdced348b08d2435eb2d4698f50ef3/Src/Newtonsoft.Json/Serialization/DefaultContractResolver.cs#L448), and it, sadly just checks valueType.IsAssignableFrom(typeof(JToken)).

**Comment by CreepyGnome** [comment]
This is old but still open, so not sure if this will help you or maybe someone else who finds this issue. If you don't want to have a direct dependency on JToken you can also have a property with the type IDictionary<string, object> that is decorated with the JsonExtensionDataAttribute and the object will still be a JToken but POCO/DTO will not have to have a Newtonsoft dependency on it and the serialization and deserialization will still function.

**Comment by ChadNedzlek** [comment]
In my case I needed to read the values (I was trying to have a more useful error message to log/display to users saying that they'd passed me unsupported properties).  I initially used IDictionary<string,object> but then was perplexed as to how to read the values I had in there before I realized they were JTokens.
